Add unlocked blackboard

veerajsk opened the pull request:
Wrike task: https://www.wrike.com/open.htm?id=1062436608

- Add UnlockedBlackboard class that provides no mutex protection
- Hide internal blackboard implementation classes in an internal namespace

veerajsk committed:
Add unlocked blackboard

- Add UnlockedBlackboard class that provides no mutex protection
- Hide internal blackboard implementation classes in an internal
namespace

Dale-Koenig commented on supplementary/alica_ros1/alica_tests/src/test/test_blackboard.cpp:
```diff
@@ -150,49 +148,49 @@ TEST_F(TestBlackboard, testWithoutPlan)
-    EXPECT_THROW((checkMapping<bool, uint64_t>(BBType::BOOL, true, BBType::UINT64, 1u)), BlackboardException);
-    EXPECT_THROW((checkMapping<bool, double>(BBType::BOOL, true, BBType::DOUBLE, 1.0)), BlackboardException);
-    EXPECT_THROW((checkMapping<bool, std::string>(BBType::BOOL, true, BBType::STRING, "true")), BlackboardException);
+    EXPECT_THROW((checkMapping<bool, int64_t>(BBType::BOOL, true, BBType::INT64, 1)), internal::BlackboardException);
```
I don't think the exception should be in internal namespace

bjoernschroeder replied:
I also think the exception should not be internal, because it is part of how a user of the blackboard interacts with the blackboard, by receiving and catching these exceptions if something goes wrong.

veerajsk replied:
yeah right, I'll change it

veerajsk committed:
Move BlackboardException outside the internal namespace

ntfshard commented on alica_engine/include/engine/blackboard/Blackboard.h:
```diff
@@ -333,12 +334,12 @@ class Blackboard
-    BlackboardImpl& impl() { return _impl; }
-    const BlackboardImpl& impl() const { return _impl; }
+    // TODO: once these APIs are removed from user code, make these methods private so that only alica internal's can use them
+    [[deprecated("Use UnlockedBlackboard instead")]] internal::BlackboardImpl& impl() { return _impl; }
```
Build log will be overfilled with this warnings, correct me if I wrong.

Dale-Koenig replied:
@veerajsk Please add a new method (private, not deprecated) that we use when constructing unlocked blackboard so that we don't get compiler warnings from normal use.  Currently even building alica will raise warnings because this function is used internally

Dale-Koenig replied:
Or internal things just use _impl directly so we can remove this function entirely instead of making it private

veerajsk replied:
Idea was to use UnlockedBlackboard eveywhere & deprecate this method until then. But yes, this is generating a bunch of warnings even in the engine. I'll work on a fix

veerajsk replied:
I switched to using _impl directly for alica internals & tests. Had to declare a bunch of friends for it though. However, seems to be the best solution

Dale-Koenig replied:
Maybe tests can just lock the blackboard?  Performance doesn't matter that much for tests and Locking really isn't that important to avoid anyway imo

ntfshard replied:
Still the motivation is unclear for me for this change. Seems code became worse after this change

veerajsk replied:
@Dale-Koenig the access to impl is only required for accessing the `map()` method, which is not exposed to the user. Otherwise the tests are mostly using the Locked blackboard methods

@ntfshard some context here: User code was using the impl() method to access the blackboard without a lock, which is bad, since user code should not be accessing internal implementation classes. However, as explained here https://github.com/rapyuta-robotics/alica/pull/425#issuecomment-1436461189 we need unlocked access to blackboard in many cases. Therefore this PR adds a UnlockedBlackboard class which the user code will now use instead of the impl() method. Once user code is change this deprecated method will be removed. I hope it is clear now?

ntfshard commented on supplementary/alica_ros1/alica_tests/src/test/test_blackboard.cpp:
```diff
@@ -7,8 +7,6 @@
 
 namespace alica
 {
-namespace
```
hope this change won't break ODR

veerajsk replied:
its working fine

ntfshard replied:
ODR violation can work randomly technically

veerajsk replied:
yes, but ODR will mostly occur if anonymous namespaces are used in headers. Here it should not matter, because this is a cpp file with mostly tests & test fixtures. Only 3 classes are defined here & all of them are exclusive to this cpp file, so there should be no ODR violations

ntfshard replied:
I told about types definitions. It's much more unpleasant problem, here I prepared a simple example https://github.com/ntfshard/cpp-odr-violation-types-example

But here, I checked and seems it shouldn't be a problem.

veerajsk replied:
ok, thank you for the example. I'll keep this in mind

Could you please approve the PR if it is fine?

ntfshard reviewed (commented):
Why do we need this non-thread safe interface after all?

ntfshard commented:
I have a proposal for interface changes:
 - move `LockedBlackboardRW`/`LockedBlackboardRO` classes inside `Blackboard` class
 - make methods for construct `LockedBlackboardRW`/`LockedBlackboardRO`  objects

usage will be:
`ac->editGlobalBlackboard().lockedBlackboardRW().set("worldmodel", std::make_shared<alicaTests::TestWorldModel>());`
and we don't need this unsafe functions

veerajsk commented:
> I have a proposal for interface changes:
> 
> * move `LockedBlackboardRW`/`LockedBlackboardRO` classes inside `Blackboard` class
> * make methods for construct `LockedBlackboardRW`/`LockedBlackboardRO`  objects
> 
> usage will be: `ac->editGlobalBlackboard().lockedBlackboardRW().set("worldmodel", std::make_shared<alicaTests::TestWorldModel>());` and we don't need this unsafe functions

I'm not sure we should make a breaking interface change for locked blackboards especially if it does not add/fix something. As for unsafe access, actually we do need it in several legitimate cases where it is guaranteed that locks are not required to access blackboard keys eg. protected keys almost never need to be locked unless the blackboard is inherited.

veerajsk committed:
Fix access to blackboard internals

Prevents warnings on accessing internals i.e. BlackboardImpl class
in alica & alica tests

veerajsk committed:
Merge branch 'devel' into vsk_unlocked_bb

veerajsk committed:
Merge branch 'devel' into vsk_unlocked_bb

veerajsk committed:
Merge branch 'devel' into vsk_unlocked_bb

veerajsk committed:
Merge branch 'devel' into vsk_unlocked_bb

ntfshard commented:
> I'm not sure we should make a breaking interface change for locked blackboards especially if it does not add/fix something. As for unsafe access, actually we do need it in several legitimate cases where it is guaranteed that locks are not required to access blackboard keys eg. protected keys almost never need to be locked unless the blackboard is inherited.

Do you have a performance measurements which can show that this new interface improves performance?

veerajsk commented:
> > I'm not sure we should make a breaking interface change for locked blackboards especially if it does not add/fix something. As for unsafe access, actually we do need it in several legitimate cases where it is guaranteed that locks are not required to access blackboard keys eg. protected keys almost never need to be locked unless the blackboard is inherited.
> 
> Do you have a performance measurements which can show that this new interface improves performance?

nope, I don't have any performance measurements

bjoernschroeder reviewed (approved):
Would be great to have one test using the new unlocked bb, but the implementation looks good

veerajsk committed:
Add unit test for unlocked blackboard

veerajsk committed:
Merge remote-tracking branch 'origin/vsk_unlocked_bb' into vsk_unlocked_bb
